src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcTransport.cs

```diff
@@ -0,0 +1,418 @@
+            }
+            else if (stream is PipeStream pipeStream)
+            {
+                Debug.Assert(RuntimeInformation.IsOSPlatform(OSPlatform.Windows), "Pipe stream should only be used on Windows.");
+
+                // PeekNamedPipe will return false if the pipe is disconnected/broken.
+                return NativeMethods.PeekNamedPipe(
+                    pipeStream.SafePipeHandle,
+                    null,
+                    0,
+                    IntPtr.Zero,
+                    IntPtr.Zero,
+                    IntPtr.Zero);
+            }
+
+            return false;
+        }
+
+        private void RegisterTarget(StreamTarget target)
+        {
+            lock (_targets)
+            {
+                // Allow transport specific implementation to refresh
+                // the stream before possibly consuming it.
+                if (null == _stream)
+                {
+                    _stream = RefreshStream();
+                }
+
+                // If there is no current stream, add the target to the queue;
+                // it will be fulfilled at some point in the future when streams
+                // are made available. Otherwise, provide the stream to the target
+                // synchronously; set the current stream to null to signify that
+                // there is no current stream available.
+                if (null == _stream)
+                {
+                    _targets.Enqueue(target);
+                }
+                else if (target.SetStream(_stream))
+                {
+                    _stream = null;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Base class for providing streams to callers.
+        /// </summary>
+        private abstract class StreamTarget : IDisposable
+        {
+            public void Dispose()
+            {
+                if (!IsDisposed)
+                {
+                    Dispose(disposing: true);
+
+                    IsDisposed = true;
+                }
+            }
+
+            protected abstract void Dispose(bool disposing);
+
+            public abstract bool SetStream(Stream stream);
+
+            public bool IsDisposed { get; private set; }
+        }
+
+        /// <summary>
+        /// Class allowing for synchronously waiting on the availability of a stream.
+        /// </summary>
+        private class StreamEventTarget : StreamTarget
+        {
+            private readonly ManualResetEvent _streamEvent = new ManualResetEvent(false);
+
+            protected override void Dispose(bool disposing)
+            {
+                if (disposing)
+                {
+                    _streamEvent.Dispose();
+                }
+            }
+
+            public override bool SetStream(Stream stream)
+            {
+                if (IsDisposed)
+                {
+                    return false;
+                }
+
+                _streamEvent.Set();
```

Comment:
Did you want to assign the stream first and then call .Set()? It seems like a race condition right now (although a very unlikely one)